### PR TITLE
Refactor operator dispatch framework across different Tensors.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77707

Refactor to clean up the following pieces:

1) Consolidate decorators to use a common way to look up operator tables.
2) Move a bunch of utilities to `op_registry_utils` and `common_op_utils` and
reuse them across ShardedTensor, ReplicatedTensor and PartialTensor.

Differential Revision: [D36465639](https://our.internmc.facebook.com/intern/diff/D36465639/)

### CODE REVIEW

File: torch/distributed/_shard/partial_tensor.py [hunk 1]
```diff
@@ -1,27 +1,24 @@
...
 from torch.distributed import distributed_c10d
-from torch.distributed._shard.sharded_tensor.api import ShardedTensor
 from torch.distributed.nn.functional import (
     reduce_scatter,
 )
+from torch.distributed._shard.common_op_utils import _register_default_op
+from torch.distributed._shard.op_registry_utils import _decorator_func
+from torch.utils._pytree import tree_map
+
+if TYPE_CHECKING:
```
Comment (wanchaol) [inline torch/distributed/_shard/partial_tensor.py:15]: it looks like we already do the conditional imports in the `reshard` function, maybe we don't need this import?

**Review by wanchaol — APPROVED** [review]
looks good! have one minor comment

**Review by fduwjj — APPROVED** [review]
lgtm

File: torch/distributed/_shard/partial_tensor.py (hunk 1 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/partial_tensor.py:15]: The "ShardedTensor" return annotation for `reshard` needs this logic.

**Pull request closed without merging**